dev/ring-buffers.lisp

```diff
@@ -9,48 +9,119 @@
+                         indexes)))
+    (svref (contents container)
+           (mod (first indexes) (total-size container)))))
 
-(defmethod item-at! ((container ring-buffer) value &rest indexes)
+(defmethod item-at! ((container ring-buffer-reverse) value &rest indexes)
   (declare (dynamic-extent indexes))
-  (setf (svref (contents container)
-                (mod (first indexes) (total-size container)))
-         value))
+  (let ((indexes (mapcar #'(lambda (index) (fifo-index container index))
+                         indexes)))
+    (setf (svref (contents container)
+                 (mod (first indexes) (total-size container)))
+          value)))
 
+(defmethod item-at! ((container ring-buffer-reverse) value &rest indexes)
+  (declare (dynamic-extent indexes))
+  (let ((indexes (mapcar #'(lambda (index) (lifo-index container index))
+                         indexes)))
+    (setf (svref (contents container)
+                 (mod (first indexes) (total-size container)))
+          value)))
+
+(defmethod lifo-index ((container ring-buffer-reverse) index)
+  "Return index converted to internal LIFO index, where items are ordered from newest to oldest."
+  (mod (1- (+ (buffer-start container)
+              (- (buffer-end container) (buffer-start container) index)))
+       (total-size container)))
+
+(defmethod fifo-index ((container ring-buffer) index)
+  "Return index converted to internal FIFO index, where items are ordered from oldest to newest."
+  (mod (+ (buffer-start container) index)
+       (total-size container)))
+
+(defmethod recent-list ((container ring-buffer-reverse))
+  "Return list of items ordered by most recent."
+  (loop for index from 0 below (- (buffer-end container) (buffer-start container))
+        for item = (item-at container index)
+        when item collect item))
 
 (defmethod increment-end ((container ring-buffer))
-   (with-slots (buffer-end buffer-start) container
-     (when (and (>= buffer-end (total-size container))
-                (= (mod buffer-end (total-size container)) buffer-start))
-       (incf buffer-start))
-     (incf buffer-end)))
+  (with-slots (buffer-end buffer-start) container
+    (when (and (>= buffer-end (total-size container))
+               (= (mod buffer-end (total-size container)) buffer-start))
+      (incf buffer-start))
+    (incf buffer-end)))
+
+(defmethod delete-item-at ((container ring-buffer) &rest indexes)
+  "Delete item using FIFO index for ring-buffer."
+  (delete-item-internal container (mapcar #'(lambda (x) (fifo-index container x)) indexes)))
+
+(defmethod delete-item-at ((container ring-buffer-reverse) &rest indexes)
+  "Delete item using LIFO index for ring-buffer-reverse.."
+  (delete-item-internal container (mapcar #'(lambda (x) (lifo-index container x)) indexes)))
+
+(defmethod delete-item-internal ((container ring-buffer) &rest indexes)
+  "Delete item using internal index (not meant to be used directly)."
```

Comment:
Prefix the name with `%` to mark it as "internal".  No need to mention "internal" then.